src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1289,6 +1286,27 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
         }
 
         /// <summary>
+        /// Adds a placeholder attribute to the <paramref name="tagBuilder" />.
+        /// </summary>
+        /// <param name="viewContext">A <see cref="ViewContext"/> instance for the current scope.</param>
+        /// <param name="tagBuilder">A <see cref="TagBuilder"/> instance.</param>
+        /// <param name="modelExplorer">The <see cref="ModelExplorer"/> for the <paramref name="expression"/>.</param>
+        /// <param name="expression">Expression name, relative to the current model.</param>
+        protected virtual void AddPlaceholderAttribute(
+            ViewContext viewContext,
```

Comment:
Suggest passing in just the `ViewDataDictionary` unless you feel an override is likely to use other `ViewContext` properties. Consistent parameters with `AddValidationAttributes()` isn't a goal.